File: MsgServerClientSdk/SdkMsgClientAndroid/sdkmsgclient/src/main/java/org/dync/teameeting/sdkmsgclient/msgs/MSMessageManager.java
package org.dync.teameeting.sdkmsgclient.msgs;

/**
 * Created by hp on 7/10/16.
 */
public class MSMessageManager {
    /**
     *  add message delegate
     *
     *  params:
     *      txtDelegate: the message delegate for callback to set
     *
     *  return:
     *      void
     */
    public void addDelegate(MSSubMessageDelegate txtDelegate) {
        MsgClient.getInstance().MCSetTxtMsgDelegate(txtDelegate);
    }

    /**
     *  remove message delegate
     *
     *  params:
     *      txtDelegate: the message delegate to delete
     *
     *  return:
     *      void
     */
    public void delDelegate(MSSubMessageDelegate txtDelegate) {
        MsgClient.getInstance().MCSetTxtMsgDelegate(null);
    }

    /**
     *  sync message when first login or refresh
     *  do not call this function in an loooooop
     *
     *  params:
     *
     *  return:
     *      void
     */
    public void syncMsg() {
        MsgClient.getInstance().MCSyncMsg();
    }

    /**
     * this will sync msg to db
     * when app will exit or crash ot leave view
     * MUST invoke this to save data
     * <p/>
     * params:
     * <p/>
     * return:
     * void
     */
    public void sync2Db() {
        MsgClient.getInstance().MCSync2Db();
    }

    /**
     *  you send message in grpId
     *
     *  params:
     *      txtMsg: the message you will send
     *      GroupId and Content and Push need to be setted
     *      GroupId: the group id you are in
     *      Content: the content you send
     *      Push: 1, the msg will be pushed if offline, 0, not push
     *
     *  return:
     *      msgid: the id of this msg
     */
    public String sendTxtMsg(MSSubMessage.MSTxtMessage txtMsg) {
            if (null==txtMsg || null == txtMsg.getGroupId() || null == txtMsg.getContent()) return null;
        if (txtMsg.getGroupId().length()==0 || txtMsg.getContent().length()==0) return null;
        return MsgClient.getInstance().MCSendTxtMsg(txtMsg);
    }

    /**
     *  you send content to usres in grpId
     *
     *  params:
     *      txtMsg: the message you will send
     *      GroupId and Content and Push need to be setted
     *      GroupId: the group id you are in
     *      Content: the message you send
     *      Push: 1, the msg will be pushed if offline, 0, not push
     *
     *  return:
     *      msgid: the id of this msg
     */
    public String sendTxtMsgTos(MSSubMessage.MSTxtMessage txtMsg, String[] userIds) {
            if (null==txtMsg || null == txtMsg.getGroupId() || null == txtMsg.getContent() || null == userIds) return null;
        if (txtMsg.getGroupId().length()==0 || txtMsg.getContent().length()==0 || userIds.length==0) return null;
        return MsgClient.getInstance().MCSendTxtMsgTos(txtMsg, userIds);
    }

    /**
     *  you send content to userId
     *
     *  params:
     *      txtMsg: the message you will send
     *      Content and ToId and Push need to be setted
     *      Content: the message you send
     *      ToId: the user id you will send to
     *      Push: 1, the msg will be pushed if offline, 0, not push
     *
     *  return:
     *      msgid: the id of this msg
     */
    public String sendTxtMsgToUser(MSSubMessage.MSTxtMessage txtMsg) {
            if (null==txtMsg || null == txtMsg.getToId() || null == txtMsg.getContent()) return null;
        if (txtMsg.getToId().length()==0 || txtMsg.getContent().length()==0) return null;
        return MsgClient.getInstance().MCSendTxtMsgToUsr(txtMsg);
    }

    /**
     *  you send content to userIds
     *
     *  params:
     *      txtMsg: the message you will send
     *      Content and Push need to be setted
     *      Content: the message you send
     *      Push: 1, the msg will be pushed if offline, 0, not push
     *      users: the users ids you will send to
     *
     *  return:
     *      msgid: the id of this msg
     */
    public String sendTxtMsgToUsers(MSSubMessage.MSTxtMessage txtMsg, String[] usrIds) {
            if (null==txtMsg || null == txtMsg.getContent() || null == usrIds) return null;
        if (txtMsg.getContent().length()==0 || usrIds.length==0) return null;
        return MsgClient.getInstance().MCSendTxtMsgToUsrs(txtMsg, usrIds);
    }

    /**
     *  hostId begin to live in grpId
     *
     *  params:
     *      txtMsg: the message you will send
     *      GroupId and ToId and Flag and Push need to be setted
     *      GroupId: the group id you are in
     *      ToId: the user you want to forbidden talk in this group
     *      Flag: 1, forbidden user, 0, cancel forbidden
     *      Push: 1, the msg will be pushed if offline, 0, not push
     *
     *  return:
     *      msgid: the id of this msg
     */
    public String sendNotifyLive(MSSubMessage.MSLivMessage livMsg) {
            if (null==livMsg || null==livMsg.getGroupId()) return null;
        if (livMsg.getGroupId().length()==0) return null;
        return MsgClient.getInstance().MCNotifyLive(livMsg);
    }

    /**
     *  you send red-envelope to hostId in grpId
     *
     *  params:
     *      txtMsg: the message you will send
     *      GroupId and ToId and and Push and Cash and Wishcont need to be setted
     *      GroupId: the group id you are in
     *      ToId: the user you want to forbidden talk in this group
     *      Push: 1, the msg will be pushed if offline, 0, not push
     *      Cash: the cash mount you will send to host
     *      Wishcont: the wish message you want say to host
     *
     *  return:
     *      msgid: the id of this msg
     */
    public String sendNotifyRedEnvelope(MSSubMessage.MSRenMessage renMsg) {
            if (null==renMsg || null == renMsg.getGroupId() || null == renMsg.getToId() || null == renMsg.getCash() || null == renMsg.getWishcont()) return null;
        if (renMsg.getGroupId().length()==0 || renMsg.getToId().length()==0 || renMsg.getCash().length()==0 || renMsg.getWishcont().length()==0) return null;
        return MsgClient.getInstance().MCNotifyRedEnvelope(renMsg);
    }

    /**
     *  userId was push to blacklist in grpId
     *
     *  params:
     *      txtMsg: the message you will send
     *      GroupId and ToId and Flag and Push need to be setted
     *      GroupId: the group id you are in
     *      ToId: the user you want to forbidden talk in this group
     *      Flag: 1, forbidden user, 0, cancel forbidden
     *      Push: 1, the msg will be pushed if offline, 0, not push
     *      notifys: the users you want to notify, want them to know, usually group managers or group owner
     *
     *  return:
     *      msgid: the id of this msg
     */
    public String sendNotifyBlacklist(MSSubMessage.MSBlkMessage blkMsg, String[] notifys) {
            if (null==blkMsg || null == blkMsg.getGroupId() || null == blkMsg.getToId()) return null;
        if (blkMsg.getGroupId().length()==0 || blkMsg.getToId().length()==0) return null;
        return MsgClient.getInstance().MCNotifyBlacklist(blkMsg, notifys);
    }

    /**
     *  userId was forbidden in grpId
     *
     *  params:
     *      txtMsg: the message you will send
     *      GroupId and ToId and Flag and Push need to be setted
     *      GroupId: the group id you are in
     *      ToId: the user you want to forbidden talk in this group
     *      Flag: 1, forbidden user, 0, cancel forbidden
     *      Push: 1, the msg will be pushed if offline, 0, not push
     *      notifys: the users you want to notify, want them to know, usually group managers or group owner
     *
     *  return:
     *      msgid: the id of this msg
     */
    public String sendNotifyForbidden(MSSubMessage.MSFbdMessage fbdMsg, String[] notifys) {
            if (null==fbdMsg || null == fbdMsg.getGroupId() || null == fbdMsg.getToId()) return null;
        if (fbdMsg.getGroupId().length()==0 || fbdMsg.getToId().length()==0) return null;
        return MsgClient.getInstance().MCNotifyForbidden(fbdMsg, notifys);
    }

    /**
     *  userId in grpId was setted to be manager, this message will notify all members
     *
     *  params:
     *      txtMsg: the message you will send
     *      GroupId and ToId and Flag and Push need to be setted
     *      GroupId: the group id you are in
     *      ToId: the user you want to set to be manager
     *      Flag: 1, set user mgr, 0, set user not mgr
     *      Push: 1, the msg will be pushed if offline, 0, not push
     *
     *  return:
     *      msgid: the id of this msg
     */
    public String sendNotifySettedMgr(MSSubMessage.MSMgrMessage mgrMsg) {
            if (null==mgrMsg || null == mgrMsg.getGroupId() || null == mgrMsg.getToId()) return null;
        if (mgrMsg.getGroupId().length()==0 || mgrMsg.getToId().length()==0) return null;
        String[] notifys = {};
        return MsgClient.getInstance().MCNotifySettedMgr(mgrMsg, notifys);
    }
}

File: MsgServer/MsgServerClient/ClientCommon/MSMessageManager.cpp
//
//  MSMessageManager.m
//  SdkMsgClientIos
//
//  Created by hp on 6/29/16.
//  Copyright © 2016 Dync. All rights reserved.
//

#include "MSMessageManager.h"
#include "MsgClient.h"
#include "rtklog.h"

void MSMessageManager::AddDelegate(const MSSubMessageDelegate* subDelegate)
{
    MsgClient::Instance().MCSetTxtMsgDelegate(subDelegate);
}

void MSMessageManager::DelDelegate(const MSSubMessageDelegate* subDelegate)
{
    MsgClient::Instance().MCSetTxtMsgDelegate(nullptr);
}

void MSMessageManager::SyncMsg()
{
    MsgClient::Instance().MCSyncMsg();
}

void MSMessageManager::Sync2Db()
{
    MsgClient::Instance().MCSync2Db();
}

const char* MSMessageManager::SendTxtMsg(MSTxtMessage* txtMsg)
{
    if (nullptr==txtMsg)
    {
        LI("sendTxtMsg txtMsg is null\n");
        return nullptr;
    }
    LI("send grpid:%s, content:%s\n", txtMsg->GetGroupId().c_str(), txtMsg->GetContent().c_str());
    if ((0==txtMsg->GetGroupId().length()) || 0==txtMsg->GetContent().length())
    {
        LI("sendTxtMsg params is null\n");
        return nullptr;
    }
    std::string msgid("");
    MsgClient::Instance().MCSendTxtMsg(msgid, txtMsg);
    return strdup(msgid.c_str());
}

const char* MSMessageManager::SendTxtMsgTos(MSTxtMessage* txtMsg, std::vector<std::string> users)
{
    if (nullptr==txtMsg) return nullptr;
    if (0==txtMsg->GetGroupId().length() || 0==txtMsg->GetContent().length() || 0==users.size()) return nullptr;
    std::string msgid("");
    std::vector<std::string> vusers;
    for (std::string name : users) {
        vusers.push_back(name);
    }
    MsgClient::Instance().MCSendTxtMsgTos(msgid, txtMsg, vusers);
    return strdup(msgid.c_str());
}

const char* MSMessageManager::SendTxtMsgToUser(MSTxtMessage* txtMsg)
{
    if (nullptr==txtMsg) return nullptr;
    if ((0==txtMsg->GetToId().length())||(txtMsg->GetContent().length()==0)) return nullptr;
    std::string msgid("");
    MsgClient::Instance().MCSendTxtMsgToUsr(msgid, txtMsg);
    return strdup(msgid.c_str());
}

const char* MSMessageManager::SendTxtMsgToUsers(MSTxtMessage* txtMsg, std::vector<std::string> users)
{
    if (nullptr==txtMsg) return nullptr;
    if ((0==txtMsg->GetContent().length())||(0==users.size())) return nullptr;
    std::string msgid("");
    std::vector<std::string> vusers;
    for (std::string name : users) {
        vusers.push_back(name);
    }
    MsgClient::Instance().MCSendTxtMsgToUsrs(msgid, txtMsg, vusers);
    return strdup(msgid.c_str());
}

const char* MSMessageManager::SendNotifyLive(MSLivMessage* livMsg)
{
    if (nullptr==livMsg) return nullptr;
    if ((0==livMsg->GetGroupId().length())) return nullptr;
    std::string msgid("");
    MsgClient::Instance().MCNotifyLive(msgid, livMsg);
    return strdup(msgid.c_str());
}

const char* MSMessageManager::SendNotifyRenEnvelope(MSRenMessage* renMsg)
{
    if (nullptr==renMsg) return nullptr;
    if ((0==renMsg->GetGroupId().length())
            ||(0==renMsg->GetToId().length())
            ||(0==renMsg->GetCash().length())
            ||(0==renMsg->GetWishcont().length())) return nullptr;
    std::string msgid("");
    MsgClient::Instance().MCNotifyRedEnvelope(msgid, renMsg);
    return strdup(msgid.c_str());
}

const char* MSMessageManager::SendNotifyBlacklist(MSBlkMessage* blkMsg, std::vector<std::string> notifys)
{
    if (nullptr==blkMsg) return nullptr;
    if ((0==blkMsg->GetGroupId().length())
            ||(0==blkMsg->GetToId().length())) return nullptr;
    std::string msgid("");
    std::vector<std::string> vusers;
    if (notifys.size()>0) {
        for (std::string name : notifys) {
            vusers.push_back(name);
        }
    }
    MsgClient::Instance().MCNotifyBlacklist(msgid, blkMsg, vusers);
    return strdup(msgid.c_str());
}

const char* MSMessageManager::SendNotifyForbidden(MSFbdMessage* fbdMsg, std::vector<std::string> notifys)
{
    if (nullptr==fbdMsg) return nullptr;
    if ((0==fbdMsg->GetGroupId().length())
        ||(0==fbdMsg->GetToId().length())) return nullptr;
    std::string msgid("");
    std::vector<std::string> vusers;
    if (notifys.size()>0) {
        for (std::string name : notifys) {
            vusers.push_back(name);
        }
    }
    MsgClient::Instance().MCNotifyForbidden(msgid, fbdMsg, vusers);
    return strdup(msgid.c_str());
}

const char* MSMessageManager::SendNotifySettedMgr(MSMgrMessage* mgrMsg)
{
    if (nullptr==mgrMsg) return nullptr;
    if ((0==mgrMsg->GetGroupId().length())||(0==mgrMsg->GetToId().length())) return nullptr;
    std::string msgid("");
    std::vector<std::string> vusers;
    MsgClient::Instance().MCNotifySettedMgr(msgid, mgrMsg, vusers);
    return strdup(msgid.c_str());
}
